contracts/AelinPool.sol

```diff
@@ -385,6 +446,106 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
         emit PurchasePoolToken(msg.sender, purchaseTokenAmount);
     }
 
+    /**
+     * @dev allows anyone to become a purchaser with a qualified erc721
+     * nft in the pool depending on the scenarios
+     *
+     * Scenarios:
+     * 1. each wallet holding a qualified NFT to deposit an unlimited amount of purchase tokens
+     * 2. certain amount of purchase tokens per wallet regardless of the number of qualified NFTs held
+     * 3. certain amount of Investment tokens per qualified NFT held
+     */
+
+    function purchasePoolTokensWithNft(NftPurchaseList[] calldata _nftPurchaseList, uint256 _purchaseTokenAmount)
+        external
+        lock
+    {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+
+        uint256 maxPurchaseTokenAmount;
+
+        for (uint256 i = 0; i < _nftPurchaseList.length; i++) {
+            NftPurchaseList memory nftPurchaseList = _nftPurchaseList[i];
+            address _collectionAddress = nftPurchaseList.collectionAddress;
+            uint256[] memory _tokenIds = nftPurchaseList.tokenIds;
+
+            NftCollectionRules memory nftCollectionRules = nftCollectionDetails[_collectionAddress];
+            require(nftCollectionRules.collectionAddress == _collectionAddress, "collection not in the pool");
+
+            if (nftCollectionRules.purchaseAmountPerToken) {
+                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount * _tokenIds.length;
+            }
+
+            if (!nftCollectionRules.purchaseAmountPerToken && nftCollectionRules.purchaseAmount > 0) {
+                require(!nftWalletUsedForPurchase[_collectionAddress][msg.sender], "wallet already used for collection");
```

Comment:
we want to keep the require error string length at 32 characters or less ideally. just change it to "wallet already used for nft set"